packages/angular-cli/commands/completion.ts

```diff
@@ -1,17 +1,176 @@
   works: 'everywhere',
-  run: function() {
-    const scriptPath = path.resolve(__dirname, '..', 'utilities', 'completion.sh');
-    const scriptOutput = fs.readFileSync(scriptPath, 'utf8');
+  availableOptions: [
+    { name: 'all',   type: Boolean, default: true,  aliases: ['a'] },
+    { name: 'bash',  type: Boolean, default: false, aliases: ['b'] },
+    { name: 'zsh',   type: Boolean, default: false, aliases: ['z'] }
+  ],
+
+  run: function (commandOptions: CompletionCommandOptions) {
+    commandOptions.all = !commandOptions.bash && !commandOptions.zsh;
+
+    const commandFiles = fs.readdirSync(__dirname)
+      .filter(file => file.match(/\.ts$/) && !file.match(/\.run.ts$/))
+      .map(file => path.parse(file).name)
+      .filter(file => {
+        return commandsToIgnore.indexOf(file) < 0;
+      })
+      .map(file => file.toLowerCase());
+
+    const commandMap = commandFiles.reduce((acc: any, curr: string) => {
+      let classifiedName = stringUtils.classify(curr);
+      let defaultImport = require(`./${curr}`).default;
+
+      acc[classifiedName] = defaultImport;
+
+      return acc;
+    }, {});
+
+    let caseBlock = '';
+
+    commandFiles.forEach(cmd => {
+      const Command = lookupCommand(commandMap, cmd);
+      const com: String[] = [];
+
+      const command = new Command({
+        ui: this.ui,
+        project: this.project,
+        commands: this.commands,
+        tasks: this.tasks
+      });
+
+      optsNg.push(command.name);
+      com.push(command.name);
+
+      if (command.aliases) {
+        command.aliases.forEach((element: String) => {
+          optsNg.push(element);
+          com.push(element);
+        });
+      }
+
+      if (command.availableOptions && command.availableOptions[0]) {
+        let opts = extractOptions (command.availableOptions);
+        caseBlock = caseBlock + '    ' + com.sort().join('|') + ') opts="' + opts + '" ;;\n';
+      }
+    });
+
+    caseBlock = 'ng|help) opts="' + optsNg.sort().join(' ') + '" ;;\n' +
+                caseBlock +
+                '    *) opts="" ;;';
+
+    console.log(`###-begin-ng-completion###
+#
+
+# ng command completion script
+#   This command supports 3 cases.
+#   1. (Default case) It prints a common completion initialisation for both Bash and Zsh.
+#      It is the result of either calling "ng completion" or "ng completion -a".
+#   2. Produce Bash-only completion: "ng completion -b" or "ng completion --bash".
+#   3. Produce Zsh-only completion: "ng completion -z" or "ng completion --zsh".
+#
+# Installation: ng completion -b 1>> ~/.bashrc
+#           or  ng completion -z 1>> ~/.zshrc
+#
+`);
+
+    if (commandOptions.all && !commandOptions.bash) {
+      console.log('if test ".$(type -t complete 2>/dev/null || true)" = ".builtin"; then');
+    }
+
+    if (commandOptions.all || commandOptions.bash) {
+      console.log(`_ng_completion() {
+  local cword pword opts
```

Comment:
Same here; use `stripIndent` so we keep the indentation in the code (I don't care much about the generated files).